src/appleseed/renderer/kernel/shading/shadingengine.cpp

```diff
@@ -105,17 +106,40 @@ void ShadingEngine::shade_hit_point(
     // Retrieve the material of the intersected surface.
     const Material* material = shading_point.get_material();
 
-    // Apply OSL transparency if needed.
-    if (material &&
-        material->get_render_data().m_shader_group &&
-        material->get_render_data().m_shader_group->has_transparency())
+    if (material != nullptr && material->get_render_data().m_shader_group)
     {
-        Alpha alpha;
-        shading_context.execute_osl_transparency(
-            *material->get_render_data().m_shader_group,
-            shading_point,
-            alpha);
-        shading_result.m_main.a *= alpha[0];
+        const ShaderGroup* sg = material->get_render_data().m_shader_group;
+
+        const bool has_matte = sg->has_matte();
+
+        if (sg->has_transparency() || has_matte)
+        {
+            shading_context.execute_osl_transparency_and_matte(
+                *material->get_render_data().m_shader_group,
```

Comment:
This is just `sg`.

With this change, the statement could be written on a single line.